src/meta_ops.rs

```diff
@@ -396,3 +400,296 @@ pub fn equal<'gc>(
         (Value::UserData(_), _) => Value::Boolean(false).into(),
     })
 }
+
+pub fn add<'gc>(
+    ctx: Context<'gc>,
+    lhs: Value<'gc>,
+    rhs: Value<'gc>,
+) -> Result<MetaResult<'gc, 2>, RuntimeError> {
+    arithmetic_meta_op(
+        ctx,
+        lhs,
+        rhs,
+        MetaMethod::Add,
+        "integer or table with __add or userdata with __add",
+        |x, y| raw_ops::add(x, y).ok_or(crate::thread::BinaryOperatorError::Add),
+    )
+}
+
+pub fn subtract<'gc>(
+    ctx: Context<'gc>,
+    lhs: Value<'gc>,
+    rhs: Value<'gc>,
+) -> Result<MetaResult<'gc, 2>, RuntimeError> {
+    arithmetic_meta_op(
+        ctx,
+        lhs,
+        rhs,
+        MetaMethod::Sub,
+        "integer or table with __sub or userdata with __sub",
+        |x, y| raw_ops::subtract(x, y).ok_or(crate::thread::BinaryOperatorError::Subtract),
+    )
+}
+
+//TODO figure out how to infer `expected_message`, `raw_op` from `m`.
+pub fn arithmetic_meta_op<'gc>(
+    ctx: Context<'gc>,
+    lhs: Value<'gc>,
+    rhs: Value<'gc>,
+    m: MetaMethod,
+    expected: &'static str,
+    raw_op: impl FnOnce(Value<'gc>, Value<'gc>) -> Result<Value<'gc>, BinaryOperatorError>,
+) -> Result<MetaResult<'gc, 2>, RuntimeError> {
+    Ok(match (lhs, rhs) {
+        (Value::Integer(a), Value::Integer(b)) => {
+            (raw_op(Value::Integer(a), Value::Integer(b))).map(MetaResult::Value)?
+        }
+        (Value::Integer(a), Value::Number(b)) => {
+            (raw_op(Value::Integer(a), Value::Number(b))).map(MetaResult::Value)?
+        }
+        (Value::Number(a), Value::Integer(b)) => {
+            (raw_op(Value::Number(a), Value::Integer(b))).map(MetaResult::Value)?
+        }
+        (Value::Number(a), Value::Number(b)) => {
+            (raw_op(Value::Number(a), Value::Number(b))).map(MetaResult::Value)?
+        }
+
+        (Value::Table(a), Value::Table(b)) => {
+            let get_sub = |t: Table<'gc>| {
+                let eq = t.metatable().map(|t| t.get(ctx, m)).unwrap_or_default();
+                if eq.is_nil() {
+                    None
+                } else {
+                    Some(eq)
+                }
+            };
+            if let Some(a_eq) = get_sub(a) {
+                MetaResult::Call(MetaCall {
+                    function: call(ctx, a_eq)?,
+                    args: [a.into(), b.into()],
+                })
+            } else if let Some(b_eq) = get_sub(b) {
+                MetaResult::Call(MetaCall {
+                    function: call(ctx, b_eq)?,
+                    args: [a.into(), b.into()],
+                })
+            } else {
+                Err(TypeError {
+                    expected,
+                    found: "tables with no __sub",
```

Comment:
If generic over metamethod, this should reflect the metamethod used, so probably a `format!("tables with no {m}")`